libvast/src/file.cpp

```diff
@@ -0,0 +1,129 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/file.hpp"
+
+#include "vast/detail/assert.hpp"
+#include "vast/detail/posix.hpp"
+#include "vast/error.hpp"
+#include "vast/path.hpp"
+
+#if VAST_POSIX
+#  include <fcntl.h>
+#  include <stdio.h>
+
+#  include <sys/stat.h>
+#  include <sys/types.h>
+#endif // VAST_POSIX
+
+namespace vast {
+
+file::file(vast::path p) : path_{std::move(p)} {
+}
+
+file::~file() {
+  // Don't close stdin/stdout implicitly.
+  if (path_ != "-")
+    close();
+}
+
+caf::expected<void> file::open(open_mode mode, bool append) {
```

Comment:
We recently had a discussion for a similar set of functions about `caf::expected<void>` vs. `caf::error`, where the latter won the vote. Should we change this to use `caf::error` as well?

Reply:
Yes, that would be the right thing to do.

Reply:
Similar to the `span`ification, I think we should tackle these as we go. These kinda stories never get priority.

Reply:
This and the other similar suggestions would be nice to have but not of the same priority as the changes introduced by the existing commits. I would prefer to focus on delivery for now.